### PR TITLE
Adding Integration Test for exceeding Overdraft balance

### PR DESCRIPTION
**Pull request by s1ddh4rthc**
`src/test/java/net/testudobank/tests/MvcControllerIntegTest.java`
- Wrote an integration test to test the edge case where a customer withdraws an amount that causes an overdraft that is within the defined limit of $1000.
- Due to the applied interest rate of 2%, though the overdraft balance is under $1000, once interest is applied, the overdraft balance can go above $1000.
- This integration test makes sure that if the `CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_AFTER_PENNIES` is above the limit (which it will be every time due to the test), the `submitWithdraw()` function is sending the user back to the home screen.
- The test also makes sure that the customer's balance stays the same since the withdraw will not go through due to the overdraft balance being above the limit.
- This test verifies that no transaction has taken place by checking if the TransactionHistory Table is empty.

### CODE REVIEW

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 1]
```diff
@@ -239,6 +239,69 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
     MvcControllerIntegTestHelpers.checkTransactionLog(customer1TransactionLog, timeWhenWithdrawRequestSent, CUSTOMER1_ID, MvcController.TRANSACTION_HISTORY_WITHDRAW_ACTION, CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES);
   }
 
+/**
+   * Verifies the case where a customer withdraws an amount that causes the
+   * customer to enter overdraft, and the overdraft balance with interest applied
+   * doesn't exceed the overdraft limit.
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:245]: Delete this first paragraph because it conflicts with the 2 paragraphs you have written below.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 2]
```diff
@@ -239,6 +239,69 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+   LocalDateTime timeWhenWithdrawRequestSent = MvcControllerIntegTestHelpers.fetchCurrentTimeAsLocalDateTimeNoMilliseconds();
+   System.out.println("Timestamp when withdraw request sent: " + timeWhenWithdrawRequestSent);
+
+
+   //verify that customer1's Overdraft balance is equal to remaining withdraw amount with interest applied
+   int CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+   int CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_WITHDRAW);
+   int CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_BEFORE_INTEREST_IN_PENNIES = CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES - CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES;
+   int CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_AFTER_INTEREST_IN_PENNIES = (int)(CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_BEFORE_INTEREST_IN_PENNIES * MvcController.INTEREST_RATE);
+   System.out.println("Expected Overdraft Balance in pennies: " + CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_AFTER_INTEREST_IN_PENNIES);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:284]: You should come up with a better variable name than `CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE` because that variable name is misleading. The actual expected overdraft balance after this transaction is $0 since we expect the withdraw to be rejected. You can try something like `CUSTOMER1_OVERDRAFT_INCREASE_AMOUNT` or something, just don't use the word `"EXPECTED"`.

Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:284]: The conventions followed by the other integ tests that you should follow is to initialize these variables AFTER sending off the POST request to the MvcController.

Also, the `timeWhenWithdrawRequestSent` initialization should come **immediately** before the withdraw request is sent, and you are messing up that pattern by having these `int` variable initializations here. Although the JVM runtime of these `int` initializations is negligible, you should avoid this pattern so that in the future you don't put some runtime-costly code between the `timeWhenWithdrawRequestSent` initialization and the withdraw request being sent because this could mess up the expected timestamps in the `TransactionHistory` and `OverdraftLogs` tables.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 3]
```diff
@@ -239,6 +239,69 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+   //Fetch customer1's data from DB
+   List<Map<String, Object>> customersTableData = jdbcTemplate.queryForList("SELECT * FROM Customers;");
+
+   //Since the request did not go through, the balance is supposed to stay the same.
+   Map<String, Object> customer1Data = customersTableData.get(0);
+   assertEquals(CUSTOMER1_BALANCE_IN_PENNIES, (int)customer1Data.get("Balance"));
+
+   //check that TransactionHistory table is empty
+   List<Map<String, Object>> transactionHistoryTableData = jdbcTemplate.queryForList("SELECT * FROM TransactionHistory;");
+   assertEquals(true, transactionHistoryTableData.isEmpty());
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:300]: Just use `assertTrue()` to make this easier to read.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 4]
```diff
@@ -239,6 +239,69 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+   String responsePage = controller.submitWithdraw(customer1WithdrawFormInputs);
+   assertEquals("welcome", responsePage);
+
+   //Fetch customer1's data from DB
+   List<Map<String, Object>> customersTableData = jdbcTemplate.queryForList("SELECT * FROM Customers;");
+
+   //Since the request did not go through, the balance is supposed to stay the same.
+   Map<String, Object> customer1Data = customersTableData.get(0);
+   assertEquals(CUSTOMER1_BALANCE_IN_PENNIES, (int)customer1Data.get("Balance"));
+
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:297]: I would also include a check that `OverdraftBalance` column in Customers table is 0 just to be safe. I've seen student code before where the balance is kept the same but the overdraft balance is increased when withdrawing.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 5]
```diff
@@ -239,6 +239,69 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+   int CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_AFTER_INTEREST_IN_PENNIES = (int)(CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_BEFORE_INTEREST_IN_PENNIES * MvcController.INTEREST_RATE);
+   System.out.println("Expected Overdraft Balance in pennies: " + CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_AFTER_INTEREST_IN_PENNIES);
+
+   
+   //Check the response when the withdraw request is submitted. This should return the user back to the home screen due to an invalid request
+   String responsePage = controller.submitWithdraw(customer1WithdrawFormInputs);
+   assertEquals("welcome", responsePage);
+
+   //Fetch customer1's data from DB
+   List<Map<String, Object>> customersTableData = jdbcTemplate.queryForList("SELECT * FROM Customers;");
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:292]: Include the `WHERE CustomerID=''` clause in this `SELECT` query.

**Review by adithyasolai — CHANGES_REQUESTED** [review]
Good job! Most of my comments below have to do with following codebase conventions and conventions in this same integ test file, as well as some clean code considerations.

Verified that integration tests pass locally.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java (hunk 1 above)
Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:245]: Thanks, I did not notice that this first paragraph was still here. I have removed it now.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java (hunk 2 above)
Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:284]: Sure, I'll change that variable name right now to something that is more accurate.

Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:284]: This makes sense. I will keep this in mind in the future as well, thanks for pointing that out to me. I have changed the code to reflect the change.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java (hunk 3 above)
Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:300]: Will do so. I originally used assertEquals because that is what was imported at the top, but I added another import statement to allow assertTrue() as well.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java (hunk 4 above)
Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:297]: Done.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 6]
```diff
@@ -275,18 +273,16 @@ public void testWithdrawOverdraftOverage() throws SQLException, ScriptException
...
    //Check the response when the withdraw request is submitted. This should return the user back to the home screen due to an invalid request
    String responsePage = controller.submitWithdraw(customer1WithdrawFormInputs);
    assertEquals("welcome", responsePage);
+   
+   //verify that customer1's Overdraft balance is equal to remaining withdraw amount with interest applied
+   int CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+   int CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_WITHDRAW);
+   int CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_BEFORE_INTEREST_IN_PENNIES = CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES - CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES;
+   int CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_AFTER_INTEREST_IN_PENNIES = (int)(CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_BEFORE_INTEREST_IN_PENNIES * MvcController.INTEREST_RATE);
+   System.out.println("Expected Overdraft Balance in pennies: " + CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_AFTER_INTEREST_IN_PENNIES);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:285]: Just remove this print statement. You use the word `"Expected"` again in the print statement. Also it should be clear from the method's name and its summary comment that the customer's balance should stay the same, so this extra documentation is not required to send to the output.

If this integ test were to crash, the terminal will display this test's name, which should indicate to the user what this integ test cast is trying to do.

I am thinking about adding a Debug Logging suite to our codebase so that we can be more lenient with outputting stuff to the terminal while testing.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 7]
```diff
@@ -239,6 +240,67 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+   * 
+   * Checks to make sure that the customer's balance stays the same as before due to a failed 
+   * withdraw request, and checks that the TransactionHistory table is empty.
+   * 
+   * @throws SQLException
+   * @throws ScriptException
+   */
+
+  @Test
+  public void testWithdrawOverdraftOverage() throws SQLException, ScriptException {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:258]: Change the name of this method to `testWithdrawOverdraftLimitExceeded()` to add more clarity. `Overage` is kind of vague.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 8]
```diff
@@ -275,18 +273,16 @@ public void testWithdrawOverdraftOverage() throws SQLException, ScriptException
...
-   
    //Check the response when the withdraw request is submitted. This should return the user back to the home screen due to an invalid request
    String responsePage = controller.submitWithdraw(customer1WithdrawFormInputs);
    assertEquals("welcome", responsePage);
+   
+   //verify that customer1's Overdraft balance is equal to remaining withdraw amount with interest applied
+   int CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+   int CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_WITHDRAW);
+   int CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_BEFORE_INTEREST_IN_PENNIES = CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES - CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES;
+   int CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_AFTER_INTEREST_IN_PENNIES = (int)(CUSTOMER1_OVERDRAFT_BALANCE_INCREASE_BEFORE_INTEREST_IN_PENNIES * MvcController.INTEREST_RATE);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:284]: I'm noticing that we apply interest rate a few times in this integ test file. I will make it an extra TODO item to make another helper method that applies interest to make the code cleaner.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 9]
```diff
@@ -239,6 +240,67 @@ public void testWithdrawTriggersOverdraft() throws SQLException, ScriptException
...
+
+   //Since the request did not go through, the balance is supposed to stay the same.
+   Map<String, Object> customer1Data = customersTableData.get(0);
+   assertEquals(CUSTOMER1_BALANCE_IN_PENNIES, (int)customer1Data.get("Balance"));
+
+   //Checks to make sure that the overdraft balance was not increased
+   assertEquals(0, (int)customer1Data.get("OverdraftBalance"));
+
+   //check that TransactionHistory table is empty
+   List<Map<String, Object>> transactionHistoryTableData = jdbcTemplate.queryForList("SELECT * FROM TransactionHistory WHERE CustomerID='';");
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:298]: I think I misguided you here. Just revert this back to `SELECT * FROM TransactionHistory;`. Also, I wanted you to include the `WHERE` clause for the balance checks above, not this check.

The problem with this code is that you are only fetching rows from TransactionHistory table where the Customer's ID is `''`, which is an empty string. Therefore, you aren't really checking if the entire table is empty in the next line.

`SELECT * FROM TransactionHistory;`, on the other hand, just blindly fetches all rows from the table, which is what you want.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java (hunk 5 above)
Comment (adithyasolai) [reply src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:288]: Just ignore this comment actually. I looked at the other integ tests, and the SQL query you used is the one that is used in other test cases. You don't need the `WHERE` clause that filters for only customer1's rows.

**Review by adithyasolai — CHANGES_REQUESTED** [review]
Just left a few nitpicks that you need to take care of before I merge.

**Review by adithyasolai — APPROVED** [review]
